plugin/storage/cassandra/spanstore/writer.go

```diff
@@ -99,29 +120,42 @@ func (s *SpanWriter) WriteSpan(span *model.Span) error {
 		ds.Logs,
 		ds.Refs,
 		ds.Process,
-		ds.ServiceName,
 	)
-	err := s.tracesTableMetrics.Exec(mainQuery, s.logger)
-	if err != nil {
+
+	if err := s.writerMetrics.traces.Exec(mainQuery, s.logger); err != nil {
 		return s.logError(ds, err, "Failed to insert span", s.logger)
 	}
-	err = s.saveServiceNameAndOperationName(ds.ServiceName, ds.OperationName)
-	if err != nil {
+	if err := s.saveServiceNameAndOperationName(ds.ServiceName, ds.OperationName); err != nil {
 		// should this be a soft failure?
 		return s.logError(ds, err, "Failed to insert service name and operation name", s.logger)
 	}
 
+	if err := s.indexByTags(span, ds); err != nil {
+		return s.logError(ds, err, "Failed to index tags", s.logger)
+	}
+
+	if err := s.indexBySericeAndOperation(span.TraceID, ds); err != nil {
+		return s.logError(ds, err, "Failed to index service and operation name", s.logger)
+	}
+
+	if err := s.indexByDuration(ds); err != nil {
+		return s.logError(ds, err, "Failed to index duration", s.logger)
+	}
+	return nil
+}
+
+func (s *SpanWriter) indexByTags(span *model.Span, ds *dbmodel.Span) error {
 	for _, v := range dbmodel.GetAllUniqueTags(span) {
 		// we should introduce retries or just ignore failures imo, retrying each individual tag insertion might be better
 		// we should consider bucketing.
 		if s.shouldIndexTag(v) {
-			insertTagQuery := s.session.Query(insertTag, ds.TraceID, ds.SpanID, v.ServiceName, v.TagKey, v.TagValue)
-			if err := s.tagIndexTableMetrics.Exec(insertTagQuery, s.logger); err != nil {
+			insertTagQuery := s.session.Query(insertTag, ds.TraceID, ds.SpanID, v.ServiceName, ds.StartTime, v.TagKey, v.TagValue)
```

Comment:
If we add an index that has both tag, service name, and operation name, (in addition to what we have now) the querying logic breaks up, what we would have to consider is that such an index will have the greatest increase in write throughput on cassandra.